### PR TITLE
Static web

### PR DESCRIPTION
**Pull request by ixv**
Revamp of /ren/tree to build statically is now complete.

Too many changes to list here, but documentation on how to use the new renderers will be forthcoming

### CODE REVIEW

File: ren/tree/kids.hoon [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      [%sub-id @t]
+      [%datapath @t]
+  ==
+=/  has-opt   ~(has in (sy opts))
+=/  named-opt  %-  malt
+  %+  turn  (skim opts |=(a=* .?(a)))
+  |=  a=*
+  ?>  ?=([@ *] a)
+  [`@tas`-.a +.a]
+=/  class-val
```
Comment (belisarius222) [inline ren/tree/kids.hoon]: These definitions could be shortened slightly using `(fall (~(get by named-opt) %class) 'list')` etc.
They should work fine as-is, though.

File: ren/tree/nav.hoon [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+=/  sorted  %+  sort  ~(tap by kid-meta)
+  |=  $:  [anom=@ta adat=(map knot cord)]
+          [bnom=@ta bdat=(map knot cord)]
+      ==
+  ^-  ?
+  =/  has-a  ~(has by adat)
+  =/  has-b  ~(has by bdat)
+  =/  got-a  ~(got by adat)
+  =/  got-b  ~(got by bdat)
+  ?:  =(%bump sort-opt)
```
Comment (belisarius222) [inline ren/tree/nav.hoon]: I still don't really understand what the bump is all about.

File: ren/tree/nav.hoon [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  ?:  &((has-a %sort) (has-b %sort))
+    %+  lte  (slav %ud (got-a %sort))
+             (slav %ud (got-b %sort))
+  ?:  (has-a %sort)
+    &
+  ?:  (has-b %sort)
+    |
+  &
+::
+=/  up-link
```
Comment (belisarius222) [inline ren/tree/nav.hoon]: These can also be shortened using `++fall`.

File: web/fora.hoon [hunk 4]
```diff
@@ -0,0 +1,18 @@
+/?  310
```
Comment (belisarius222) [inline web/fora.hoon]: All the fora files belong in a different PR. They must have been added before you added those paths in the .gitignore.

File: web/tree/util.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+function editPlanInfo(viewmode){
+var edit = document.getElementsByClassName('edit')[0];
+
+  who = document.getElementsByTagName('h6')[0];
+  loc = document.getElementById('loc');
+  if (viewmode) {
+    who.style.display = 'block';
+
+    var whoIn = document.createElement("input");
+    whoIn.placeholder = who.innerHTML || 'Sun Tzu';
```
Comment (belisarius222) [inline web/tree/util.js]: Should this still be in there? Looks a bit like a debugging artifact.

File: web/tree/util.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+  var params = {
+    appl: "fora",
+    mark: "fora-post",
+  };
+
+  window.urb.send(data, params, function(error, result){
+    window.location.href = "../../fora.last-post?by=~"+window.urb.ship;
+  });
+
```
Comment (belisarius222) [inline web/tree/util.js]: There are a few extra blank lines here.

File: web/fora.hoon (hunk 4 above)
Comment (ohAitch) [reply web/fora.hoon]: A PR to [urbit/fora](https://github.com/urbit/fora), to be precise.

File: lib/hood/write.hoon [hunk 7]
```diff
@@ -86,31 +86,47 @@
   |=  {sup/path him/ship txt/@t}  ^+  abet
   =+  pax=(welp (flop sup) /comments/(scot %da now))
   =.  txt
-    %+  rap  3  :~
-      '## `'  (scot %p him)  '`'
-      '\0a'  txt
-    ==
-  (poke--data [`%md pax] %mime / (as-octs txt))
+    %-  crip
```
[the remark this answers is not in the document]
Comment (ixv) [reply lib/hood/write.hoon]: I mean we could just put that information in the content like the old version does, but this way works better for the css and I also use the date field for the bump renderer. (I could use the file name for this in this case, but I wanted the bump renderer to be more general)

File: lib/hood/write.hoon [hunk 8]
```diff
@@ -86,31 +86,47 @@
...
+      ;>
+      {(trip txt)}
+
+      """
+  (poke--data [`%umd pax] %mime / (as-octs txt))
 ::
 ++  poke-fora-post
   |=  {sup/path him/ship hed/@t txt/@t}  ^+  abet
   =+  pax=(welp (flop sup) /posts/(cat 3 (scot %da now) '~'))
+  =+  spur=(spud (flop pax))
```
[the remark this answers is not in the document]
Comment (ixv) [reply lib/hood/write.hoon]: removed

File: mar/body.hoon [hunk 9]
```diff
@@ -0,0 +1,38 @@
+::
+::::  /hoon/body/mar
+  ::
+/?    310
+=,  html
+|%
+  ++  hedtal
```
[the remark this answers is not in the document]
Comment (ixv) [reply mar/body.hoon]: done

File: mar/body.hoon [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    =|  met/marl
+    |=  a/marl  ^-  {hed/marl tal/marl}
+    ?~  a  [~ ~]
+    ?.  ?=($h1 n.g.i.a)
```
[the remark this answers is not in the document]
Comment (ixv) [reply mar/body.hoon]: done

File: mar/body.hoon [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+      ?:  ?=($meta n.g.i.a)
+        $(a t.a, met [i.a met])
+      =+  had=$(a c.i.a)
```
[the remark this answers is not in the document]
Comment (ixv) [reply mar/body.hoon]: done

File: mar/body.hoon [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    [c.i.a (weld (flop met) t.a)]
+--
+::
+::
+=,  mimes:html
+|_  {hed/marl tal/marl}
+::
+++  grow                                                ::  convert to
+  |%
+  ++  mime
```
[the remark this answers is not in the document]
Comment (ixv) [reply mar/body.hoon]: Yeah I'm not sure why this is, mar/body is basically copied from mar/snip, but I changed this and it seems to work fine

File: mar/flem.hoon [hunk 13]
```diff
@@ -0,0 +1,23 @@
+::
+::::  /hoon/flem/mar
+  ::
+/?    310
+=,  mimes:html
+=,  html
+|%
++=  flem  [(list [knot cord]) manx]
```
[the remark this answers is not in the document]
Comment (ixv) [reply mar/flem.hoon]: included the description "elem with frontmatter" which I hope is enough, I can't think of a descriptive name other than like "front-elem" which doesn't add much in my option

File: mar/flem.hoon [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+--
+|_  own/flem
+::
+++  grow                                                ::  convert to
+  |%
+  ++  hymn  ;html:(head body:"+{+.own}")                ::  convert to %hymn
```
[the remark this answers is not in the document]
Comment (ixv) [reply mar/flem.hoon]: +{} for the manx interpolation, +.own is lark syntax to get the manx itself

File: mar/full.hoon [hunk 15]
```diff
@@ -0,0 +1,38 @@
+::
+::::  /hoon/full/mar
+  ::
+/?    310
+=,  html
+|%
+  ++  hedtal
```
[the remark this answers is not in the document]
Comment (ixv) [reply mar/full.hoon]: this file is old and shouldn't be here, removed

File: mar/last-post.hoon [hunk 16]
```diff
@@ -0,0 +1,28 @@
+::
+::::  /hoon/last-post/mar
+  ::
+/?    310
+=,  mimes:html
+=,  html
+|_  url/tape
```
[the remark this answers is not in the document]
Comment (ixv) [reply mar/last-post.hoon]: its in a renderer/mark because otherwise the only way would be to get the file url through the fora app, which would indiscriminately redirect everyone browsing the fora. I put a description in ren/last-post, I hope thats enough

File: ren/last-post.hoon [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+/?  310
+/=  gas   /$  fuel:html
+/=  meta  /:  /%/posts  
+          /^  (map @ta (map knot cord))
+          /_  /%  /tree-front/
+=/  self  (fall (~(get by qix.gas) %by) '')
+=/  filtered=(list [@ta (map knot cord)])  
+  %+  skim  ~(tap by meta)
+  |=  [pos=@ta dat=(map knot cord)]
+  ?.  (~(has by dat) %author)
```
[the remark this answers is not in the document]
Comment (ixv) [reply ren/last-post.hoon]: done

File: ren/tree/bump.hoon [hunk 18]
```diff
@@ -0,0 +1,21 @@
+::
+::::  /hoon/bump/tree/ren
+  ::
+/?  310
+/=  meta  /^  (map knot cord)
+          /&front&/!flem/
+/=  kids  /^  (map @ta (map knot cord))
+          /:  /%/comments  /_  /front/
+?:  =(0 ~(wyt by kids))
```
[the remark this answers is not in the document]
Comment (ixv) [reply ren/tree/bump.hoon]: bump returns the date of the last file in a directory, nav and kids use this as a sort key so as to sort forum threads by the latest bumped thread. 
I put an explanation at the top of the file

File: ren/tree/comments.hoon [hunk 19]
```diff
@@ -1,21 +1,17 @@
...
-  /&elem&/md/
-::
-::::
-  ::
-`(list (pair time {ship marl}))`-.-
+/?  310
+/=  thread  /:  /%/comments  /%  /tree-kids/
+|=  sup/spur
+^-  manx
+
```
[the remark this answers is not in the document]
Comment (ixv) [reply ren/tree/comments.hoon]: done

File: ren/tree/head.hoon [hunk 20]
```diff
@@ -18,18 +18,17 @@
...
-::
-|%
-++  cdnjs
-  |=(a/tape "//cdnjs.cloudflare.com/ajax/libs/{a}{?:(nomin.dbg "" ".min")}.js")
-++  maxcdn
-  |=(a/tape "//maxcdn.bootstrapcdn.com/{a}{?:(nomin.dbg "" ".min")}.js")
---
+/=  meta  /^  (map knot cord)
+          /%  /tree-front/
+=/  title  ?:  (~(has by meta) %title)
```
[the remark this answers is not in the document]
Comment (ixv) [reply ren/tree/head.hoon]: done

File: ren/tree/kids.hoon (hunk 1 above)
Comment (ixv) [reply ren/tree/kids.hoon]: done

File: ren/tree/nav.hoon (hunk 3 above)
Comment (ixv) [reply ren/tree/nav.hoon]: done

File: web/fora.hoon (hunk 4 above)
Comment (ixv) [reply web/fora.hoon]: removed it from this repo, PR to urbit/fora coming in a bit

File: web/tree/util.js (hunk 6 above)
Comment (ixv) [reply web/tree/util.js]: removed

File: web/tree/util.js (hunk 5 above)
Comment (ixv) [reply web/tree/util.js]: not sure exactly what you're talking about here but if you mean the 'Sun Tzu' thing, thats just to keep it the same as the old home page when you edit your name.

File: lib/hood/write.hoon (hunk 7 above)
[the remark this answers is not in the document]
Comment (belisarius222) [reply lib/hood/write.hoon]: Ok. Sounds reasonable.

File: mar/flem.hoon (hunk 13 above)
[the remark this answers is not in the document]
Comment (belisarius222) [reply mar/flem.hoon]: I think that's enough, since people can look up elem.

File: mar/flem.hoon (hunk 14 above)
[the remark this answers is not in the document]
Comment (belisarius222) [reply mar/flem.hoon]: ok

File: ren/tree/kids.hoon [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+=/  sub-class-val
+  (fall (~(get by named-opt) %sub-class) '')
+?>  ?=(@t sub-class-val)
+::
+=/  sub-id-val
+  (fall (~(get by named-opt) %sub-id) '')
+?>  ?=(@t sub-id-val)
+::
+=/  sort-opt
+  ?.  (~(has by named-opt) %sort)
```
Comment (belisarius222) [inline ren/tree/kids.hoon]: one more ++fall call

File: ren/tree/bump.hoon (hunk 18 above)
[the remark this answers is not in the document]
Comment (belisarius222) [reply ren/tree/bump.hoon]: I don't see new text here. Maybe it didn't get pushed?

File: web/tree/util.js (hunk 5 above)
Comment (belisarius222) [reply web/tree/util.js]: Ok, that's fine then.

**Comment by belisarius222** [comment]
The code is almost done now -- looks like we're just two more tiny edits away. Let's start thinking about how to get this into production. @keatondunsford do you have a sense for how we should integrate this code? I'm pretty sure anything depending on old ren/tree won't work anymore once we merge this.

**Comment by keatondunsford** [comment]
@ixv I think I found a bug here, in the kids renderer.

I have to pass a path into the `datapath` option that ends with a backslash `/`, or else the links break.

To reproduce, make a small directory structure:

```
test.hoon
test/
test/case.umd
```

`test.hoon`:
```
/=  kids  /%  /tree-kids/
:-  :~  title/'Test'
    ==
;>

Here's a line.

;+  (kids %title datapath/'/test/' ~)
```

case.umd
```
:-  :~  title/'Kid file'
    ==
;>

Just a `.umd` file.
```

Load up `localhost:8080/test` to see that that renders fine. Click on the `Kid file` link to see that the nav works fine.

Now remove the last backslash from the `datapath`:

```

```

Try clicking on the same `Kid file` link and notice how you get redirected to `localhost:8080/testcase` instead of `localhost:8080/test/case`.

Paths should never end in a trailing backslash, as that is a syntax error in Hoon. I think for some reason we made it so that Eyre automatically adds a trailing backslash to links, but this shouldn't be this way.

@belisarius222 @ohAitch , comments?

**Comment by keatondunsford** [comment]
Regarding the integration plan for this PR:

There are a lot of projects going on -- @belisarius222 and @ohAitch's work on Ford, @c-johnson's work on the new SPA router, @Fang-'s work on Collections, Talk/Hall subscriptions and Clay permissions, @vvisigoth, @galenwp and @jyng's on our new UI's -- that require coherent integration in a single release. We can't release these features one-by-one on master until they're all done.

So we created a new `mve` branch -- minimum viable exit, from the old world -- off of master. All pull requests related to this overarching web overhaul project should be made against this branch. I'll take on the responsibility of merging master into `mve` whenever there are new commits on master.

@ixv, can you resubmit this PR against `mve`?

**Comment by ohAitch** [comment]
Wait, what's wrong with merging our work on ford directly onto the network?
I thought we were trying to get away from the "all pulls to a development
branch that doesn't work" model

**Comment by belisarius222** [comment]
Our Ford work will go straight to the network as soon as it's ready. The
other projects mentioned depend on new ren/tree, and that'll probably be a
different release.

**Comment by ohAitch** [comment]
The SPA router can sit around scoped to a path, Clay permissions are
backwards compatible, talk subscriptions that I know of have been fixing
performance on master mostly and also shouldn't have any inherent
dependencies, collections was last touched by me and — well, is not ready
to merge certainly, but will be once it's done and just depend on the ford
work for performance. This PR iirc has unusable performance similar to
collections and thus blocked on ford: until ford is ready, there isn't much
point putting it anywhere mainline, and after ford is ready it can also go
on master and urbit.org.
The UI work in fact belongs on a branch, sure, and that branch can pull
things from master as they become useful. I feel like I'm missing something
obvious?

**Comment by belisarius222** [comment]
The UI work will depend on the mxe branch. We're going to treat the mxe
branch as a release branch for the UI work, so we'll want it to stay in a
state of "working." It won't be ready for deployment until Ford is faster
and the UIs for collections, docs, fora, web-talk, etc. have been ported to
it.

In the meantime, we won't have to have two parallel versions of ren/tree in
any branch -- I understand that's possible, but we don't need to do it. New
UI work can rely cleanly on only the new ren/tree and we'll have one
shared, public branch of UI work as a synchronization point.
